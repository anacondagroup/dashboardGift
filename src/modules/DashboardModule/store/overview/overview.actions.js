import {
  OVERVIEW_LOAD_REQUEST,
  OVERVIEW_LOAD_SUCCESS,
  OVERVIEW_LOAD_FAIL,
  OVERVIEW_DOWNLOAD_REPORT_REQUEST,
  OVERVIEW_DOWNLOAD_REPORT_SUCCESS,
  OVERVIEW_DOWNLOAD_REPORT_FAIL,
} from './overview.types';

export const overviewLoadRequest = options => ({
  type: OVERVIEW_LOAD_REQUEST,
  payload: options,
});

export const overviewLoadSuccess = payload => ({
  type: OVERVIEW_LOAD_SUCCESS,
  payload,
});

export const overviewLoadFail = payload => ({
  type: OVERVIEW_LOAD_FAIL,
  payload,
});

export const overviewDownloadReportRequest = payload => ({
  type: OVERVIEW_DOWNLOAD_REPORT_REQUEST,
  payload,
});

export const overviewDownloadReportSuccess = () => ({
  type: OVERVIEW_DOWNLOAD_REPORT_SUCCESS,
});

export const overviewDownloadReportFail = () => ({
  type: OVERVIEW_DOWNLOAD_REPORT_FAIL,
});
