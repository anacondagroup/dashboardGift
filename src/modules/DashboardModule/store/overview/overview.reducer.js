import {
  OVERVIEW_LOAD_SUCCESS,
  OVERVIEW_LOAD_REQUEST,
  OVERVIEW_LOAD_FAIL,
  OVERVIEW_DOWNLOAD_REPORT_REQUEST,
  OVERVIEW_DOWNLOAD_REPORT_SUCCESS,
  OVERVIEW_DOWNLOAD_REPORT_FAIL,
} from './overview.types';

export const initialState = {
  errors: {},
  kpi: {},
  statuses: {},
  downloadReportLink: '',
  isLoading: false,
  isReportLoading: false,
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case OVERVIEW_LOAD_REQUEST:
      return {
        ...state,
        isLoading: true,
      };
    case OVERVIEW_LOAD_SUCCESS:
      return {
        ...state,
        isLoading: false,
        statuses: action.payload.gift_statuses || {},
        kpi: action.payload.kpi,
        downloadReportLink: action.payload.report_link,
      };
    case OVERVIEW_LOAD_FAIL:
      return {
        ...state,
        isLoading: false,
        errors: action.payload,
      };
    case OVERVIEW_DOWNLOAD_REPORT_REQUEST:
      return {
        ...state,
        isReportLoading: true,
      };
    case OVERVIEW_DOWNLOAD_REPORT_SUCCESS:
    case OVERVIEW_DOWNLOAD_REPORT_FAIL:
      return {
        ...state,
        isReportLoading: false,
      };
    default:
      return state;
  }
};
