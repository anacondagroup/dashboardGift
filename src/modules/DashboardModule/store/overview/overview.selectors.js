import * as R from 'ramda';
import { createSelector } from 'reselect';
import { updateSearch } from '@alycecom/modules';
import { pickNotEmpty, renameKeys } from '@alycecom/utils';

import { makeGetTeamById } from '../../../../store/teams/teams.selectors';
import { makeGetMemberById } from '../members/members.selectors';
import { makeGetCampaignById } from '../../../../store/campaigns/campaigns.selectors';
import { createSelectors } from '../../helpers/breakdownSelectors.helpers';

const convertObjectToUrlSearch = R.curryN(2, updateSearch)('');
export const overviewApiParams = R.pick(['campaignId', 'teamId', 'dateRangeFrom', 'dateRangeTo', 'memberId']);

export const [, , , getOverviewReportIsLoading] = createSelectors(R.prop('dashboard'))('overview');

export const getApiCallQuery = R.compose(
  convertObjectToUrlSearch,
  renameKeys({
    memberId: 'team_member_id',
    campaignId: 'campaign_id',
    teamId: 'team_id',
    dateRangeFrom: 'date_range_from',
    dateRangeTo: 'date_range_to',
  }),
  pickNotEmpty,
  overviewApiParams,
);

export const getReportFileName = apiParams => state => {
  let team = null;
  let dateFrom = null;
  let dateTo = null;
  let member = null;
  let campaign = null;

  if (apiParams.teamId) {
    const { name: teamName } = makeGetTeamById(apiParams.teamId)(state);
    team = `-${teamName}-Team`;
  }

  if (apiParams.dateRangeFrom) {
    dateFrom = `-${apiParams.dateRangeFrom}`;
  }
  if (apiParams.dateRangeTo) {
    dateTo = `-${apiParams.dateRangeTo}`;
  }

  if (apiParams.campaignId) {
    const { name: campaignName } = makeGetCampaignById(apiParams.campaignId)(state);
    campaign = `-${campaignName}-Campaign`;
  }

  if (apiParams.memberId) {
    const { fullName: memberName } = makeGetMemberById(apiParams.memberId)(state);
    member = `-${memberName}`;
  }

  return `Current-gift-statuses${member || ''}${campaign || ''}${team || ''}${dateFrom || ''}${
    dateTo ? `-${dateTo}` : ''
  }.xlsx`;
};

export const overviewSelector = createSelector(
  state => state.dashboard.overview,
  overview => ({
    kpi: overview.kpi,
    statuses: overview.statuses,
    isLoading: overview.isLoading,
    total: R.compose(
      R.sum,
      R.filter(val => typeof val === 'number'),
      R.values,
    )(overview.statuses),
  }),
);
