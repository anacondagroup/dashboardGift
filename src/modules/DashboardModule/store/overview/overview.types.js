export const OVERVIEW_LOAD_REQUEST = 'OVERVIEW_LOAD_REQUEST';
export const OVERVIEW_LOAD_SUCCESS = 'OVERVIEW_LOAD_SUCCESS';
export const OVERVIEW_LOAD_FAIL = 'OVERVIEW_LOAD_FAIL';

export const OVERVIEW_DOWNLOAD_REPORT_REQUEST = 'OVERVIEW_DOWNLOAD_REPORT_REQUEST';
export const OVERVIEW_DOWNLOAD_REPORT_SUCCESS = 'OVERVIEW_DOWNLOAD_REPORT_SUCCESS';
export const OVERVIEW_DOWNLOAD_REPORT_FAIL = 'OVERVIEW_DOWNLOAD_REPORT_FAIL';
