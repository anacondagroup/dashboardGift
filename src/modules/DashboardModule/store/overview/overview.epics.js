import { ofType } from 'redux-observable';
import { catchError, map, mergeMap, debounce, withLatestFrom, tap, takeUntil } from 'rxjs/operators/index';
import { of, timer } from 'rxjs';

import { OVERVIEW_LOAD_REQUEST, OVERVIEW_DOWNLOAD_REPORT_REQUEST } from './overview.types';
import {
  overviewLoadFail,
  overviewLoadSuccess,
  overviewDownloadReportSuccess,
  overviewDownloadReportFail,
} from './overview.actions';
import { getApiCallQuery, getReportFileName } from './overview.selectors';

export const loadOverviewEpic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(OVERVIEW_LOAD_REQUEST),
    debounce(() => timer(300)),
    map(action => `/enterprise/dashboard/statistics?${getApiCallQuery(action.payload)}`),
    mergeMap(url =>
      apiService.get(url).pipe(
        map(response => overviewLoadSuccess(response)),
        catchError(response => of(overviewLoadFail(response))),
        takeUntil(action$.ofType(OVERVIEW_LOAD_REQUEST)),
      ),
    ),
  );

export const downloadReportEpic = (action$, state$, { apiService, downloadFile }) =>
  action$.pipe(
    ofType(OVERVIEW_DOWNLOAD_REPORT_REQUEST),
    map(action => [`/enterprise/report/gift_statuses?${getApiCallQuery(action.payload)}`, action]),
    mergeMap(([url, action]) =>
      apiService.getFile(url).pipe(
        withLatestFrom(state$.pipe(map(getReportFileName(action.payload)))),
        tap(([blob, filename]) => {
          downloadFile(blob, filename);
        }),
        map(() => overviewDownloadReportSuccess()),
        catchError(() => [overviewDownloadReportFail()]),
      ),
    ),
  );

export const overviewEpics = [loadOverviewEpic, downloadReportEpic];
