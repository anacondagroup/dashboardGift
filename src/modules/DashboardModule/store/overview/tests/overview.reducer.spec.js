import { reducer, initialState } from '../overview.reducer';
import { overviewLoadSuccess } from '../overview.actions';
import { overviewMock } from '../__mocks__/overview.mock';

describe('Overview reducer', () => {
  it('Should transform API data', () => {
    const actual = reducer(initialState, overviewLoadSuccess(overviewMock));

    expect(actual).toMatchSnapshot();
  });
});
