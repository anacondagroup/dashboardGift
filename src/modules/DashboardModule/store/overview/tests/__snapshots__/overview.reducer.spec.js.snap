// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview reducer Should transform API data 1`] = `
Object {
  "downloadReportLink": "https://dashboard.alyce.com/enterprise/report/gift_statuses?team_id=123&date_range=month",
  "errors": Object {},
  "isLoading": false,
  "isReportLoading": false,
  "kpi": Object {
    "conversion": Object {
      "sent_to_booked": 33,
      "view_to_booked": 49,
    },
    "invites": Object {
      "accepted": 244,
      "created": 123,
      "delivered": 700,
      "sent": 668,
      "viewed": 200,
    },
    "meetings": Object {
      "booked": 20,
    },
  },
  "statuses": Object {
    "accepted": 0,
    "accepted_and_meeting_booked": 0,
    "accepted_and_meeting_held": 0,
    "accepted_and_meeting_not_held": 0,
    "address_being_verified": 0,
    "awaiting_fulfilment": 0,
    "bounced_or_returned": 0,
    "created_and_queued_for_research": 0,
    "declined": 0,
    "delivered_but_now_viewed": 0,
    "disabled": 0,
    "expired": 0,
    "in_transit": 0,
    "invite_options_ready": 0,
    "queued_for_shipping": 0,
    "requiring_attention": 0,
    "scheduled_to_be_sent": 0,
    "viewed": 0,
  },
}
`;
